{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8b69982f",
   "metadata": {},
   "source": [
    "# Introduction to Pangenomics\n",
    "\n",
    "## What is a \"pangenome\"?\n",
    "\n",
    "The term “pangenome” was first coined by Sigaux et al. and was used to describe a public database containing an assessment of genome and transcriptome alterations in major types of tumors, tissues, and experimental models.\n",
    "\n",
    "+ Sigaux F. Génome du cancer ou de la construction des cartes d'identité moléculaire des tumeurs [Cancer genome or the development of molecular portraits of tumors]. Bull Acad Natl Med. 2000;184(7):1441-7; discussion 1448-9. French. PMID: 11261250.\n",
    "\n",
    "![Sigaux et al.](./Figures/Abstract1.png){width=80%}\n",
    "\n",
    "The term was later revitalized by Tettelin et al. to describe a microbial genome by which genes were in the core (present in all strains) and which genes were dispensable (missing from one or more of the strains).\n",
    "\n",
    "![Tettelin et al.](./Figures/Abstract2.png){width=80%}\n",
    "\n",
    "+ Tettelin, H., Masignani, V., Cieslewicz, M. J., Donati, C., Medini, D., Ward, N. L., ... & Fraser, C. M. (2005). Genome analysis of multiple pathogenic isolates of Streptococcus agalactiae: implications for the microbial “pan-genome”. Proceedings of the National Academy of Sciences, 102(39), 13950-13955.\n",
    "\n",
    "![Pangenome: https://en.wikipedia.org/wiki/Pan-genome](./Figures/Pangenome.png){width=80%}\n",
    "\n",
    "### Open vs. Closed Genomes\n",
    "\n",
    "![Open and Closed Pangenomes: https://en.wikipedia.org/wiki/Pan-genome](./Figures/ClosedvOpen.png){width=65%}\n",
    "\n",
    "### Then vs. Now\n",
    "\n",
    "![Cost per Genome: https://www.genome.gov/about-genomics/fact-sheets/DNA-Sequencing-Costs-Data](./Figures/CostGenome.png){width=90%}\n",
    "\n",
    "- Low Cost\n",
    "- High Quality Long Reads (HiFi)\n",
    "- Many reference-quality assemblies per species\n",
    "\n",
    "![Pangenome Publications: https://www.nature.com/articles/s41477-020-0733-0](./Figures/GenomePubs.png){width=75%}\n",
    "\n",
    "### \"Pangenome\" Today\n",
    "\n",
    "“Any collection of genomic sequences to be analyzed jointly or to be used as a reference. These sequences can be linked in a graph-like structure, or simply constitute sets of (aligned or unaligned) sequences.” – Computational Pangenomics Consortium\n",
    "\n",
    "https://academic.oup.com/bib/article/19/1/118/2566735\n",
    "\n",
    "### The Benefit of Pangenomes\n",
    "\n",
    "+ Removes reference bias\n",
    "  + May only represent one organism\n",
    "  + Could be a “mosaic”of individuals, i.e. doesn’t represent a coherent haplotype\n",
    "  + Allele bias\n",
    "  + Doesn’t include common variation\n",
    "+ Allow multiple assemblies to be analyzed simultaneously, i.e. efficiently\n",
    "\n",
    "###  What are pangenomes good for?\n",
    "\n",
    "+ Core vs dispensable genes:\n",
    "  + How big is the core?\n",
    "  + How big is the dispensable?\n",
    "  + How big is the pangenome?\n",
    "  + What traits are associated with the core/dispensable?\n",
    "+ Unbiased read mapping and variant calling\n",
    "+ More robust variation-trait association\n",
    "+ Visual exploration of genomic structure of population\n",
    "\n",
    "##  Computational Pangenomics\n",
    "\n",
    "“Questions about efficient data structures, algorithms and statistical methods to perform bioinformatic analyses of pan-genomes give rise to the discipline of ‘computational pan-genomics’.”\n",
    "\n",
    " ![Computational Pangenomics: https://academic.oup.com/bib/article/19/1/118/2566735](./Figures/Computational.png){width=90%}    \n",
    "\n",
    "### Pangenome Representations\n",
    "\n",
    "+ Gene sets\n",
    "+ Multiple sequence alignments\n",
    "+ K-mer sets\n",
    "+ Graphs\n",
    "  + De Bruijn graphs\n",
    "  + Haptotype graphs\n",
    "  + **Variation graphs**\n",
    "\n",
    "### Variation Graphs\n",
    "\n",
    "+ Variation forms bubbles Nodes represent sequences\n",
    "+ Chains of nodes represent contiguous sequence in one or more assemblies\n",
    "+ The sequences of nodes connected by an edge may overlap\n",
    "+ Graphs can be acyclic or cyclic\n",
    "+ Haplotypes are “threaded” through graph as paths\n",
    "\n",
    "![Pangenome Representations: https://academic.oup.com/bib/article/19/1/118/2566735](./Figures/VariationGraphA.png){width=96%}\n",
    "\n",
    "![Pangenome Representations: https://academic.oup.com/bib/article/19/1/118/2566735](./Figures/VariationGraphB.png){width=96%}\n",
    "\n",
    "### Types of Variation Graphs\n",
    "\n",
    "1. Reference Graph (vg)\n",
    "  + A reference with variants\n",
    "  + E.G. [Human reference now includes VCF with common variation](https://www.ncbi.nlm.nih.gov/genome/guide/human/)\n",
    "2. Reference Backbone; “iterative” (minigraph)\n",
    "  + Graph starts as reference and other sequences are layered\n",
    "on, i.e. variants can be relative to sequences other than the\n",
    "reference\n",
    "3. Reference-Free (Cactus and pggb)\n",
    "  + Graph is built using non-reference techniques, such as multiple sequence alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "887e6d2e",
   "metadata": {},
   "source": [
    "These are all methods used by the [Human Pangenome Reference Consortium](https://humanpangenome.org)\n",
    "\n",
    "### Mapping Reads to Variation Graphs\n",
    "\n",
    "![Genotyping Variation: https://link.springer.com/article/10.1186/s13059-020-1941-7](./Figures/ReadsToVariation.png){width=100%}\n",
    "\n",
    "## Pangenome Data Sets\n",
    "\n",
    "+ [Human Reference + Variation VCF](https://www.ncbi.nlm.nih.gov/genome/guide/human/)\n",
    "+ [Human Pangenome Reference Consortium](https://humanpangenome.org)\n",
    "+ [Zoonomia (200 mammalss) Project](https://zoonomiaproject.org/the-data/)\n",
    "+ [Maize NAM founder genomes](https://www.science.org/doi/10.1126/science.abg5289)\n",
    "+ [Yeast Population Reference Panel (YPRP)](https://yjx1217.github.io/Yeast_PacBio_2016/welcome/)\n",
    "\n",
    "### Data/Yeast Genomes:\n",
    "\n",
    "![Yeast Genomes: https://yjx1217.github.io/Yeast_PacBio_2016/welcome/](./Figures/Yeast.png){width=100%}\n",
    "\n",
    "+ 12 Mb\n",
    "+ 16 chromosomes\n",
    "+ [12 strains from Yeast Population Reference Panel (YPRP)](https://yjx1217.github.io/Yeast_PacBio_2016/welcome/)\n",
    "  + 7 Saccharomyces cerevisiae (brewer’s yeast)\n",
    "    + Includes S288C reference\n",
    "+ 5 *Saccharomyces paradoxus* (wild yeast)\n",
    "\n",
    "+ [Manuscript](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2659681/)\n",
    "+ Software (LRSDAY)\n",
    "  + [Manuscript](https://www.nature.com/articles/nprot.2018.025)\n",
    "  + [GitHub](https://github.com/yjx1217/LRSDAY)\n",
    "\n",
    "### Yeast Assemblies\n",
    "\n",
    "+ [YPRP: 12 Yeast PacBio Assemblies (Chromosome level)](https://yjx1217.github.io/Yeast_PacBio_2016/data/)\n",
    "  + ~100-200x PacBio sequencing reads\n",
    "  + HGAP + Quiver polishing\n",
    "  + ~200-500x Illumina (Pilon correction)\n",
    "  + Manual curation\n",
    "  + Annotation\n",
    "\n",
    "### SK1 Illumina Reads\n",
    "\n",
    "SK1 is the most distant from S288C\n",
    "\n",
    "![Yeast Genomes: https://yjx1217.github.io/Yeast_PacBio_2016/welcome/](./Figures/YeastB.png){width=100%}\n",
    "\n",
    "### CUP1 Gene\n",
    "\n",
    "![Structrual Rearrangements: https://www.nature.com/articles/ng.3847](./Figures/StructuralRearrangements.png){width=100%}\n",
    "\n",
    "+ [CUP1](https://www.yeastgenome.org/locus/S000001095) - A gene involved in heavy metal (copper) tolerance with copy-number variation (CNV) in population.\n",
    "+ [YHR054C](https://www.yeastgenome.org/locus/S000001096) - Putative protein of unknown function.\n",
    "\n",
    "### We Changed the Names\n",
    "\n",
    "+ YPRP FASTA files only contain chromosome names\n",
    "+ We prefixed every chromosome with its assembly name and a \".\" delineator\n",
    "  + e.g. S288C.chrVIII\n",
    "+ [Pangenome Sequence Naming Specification](https://github.com/pangenome/PanSN-spec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88939a3a",
   "metadata": {
    "tags": [
     "remove_cell"
    ]
   },
   "outputs": [],
   "source": [
    "# automatically create a bib database for R packages\n",
    "knitr::write_bib(c(\n",
    "  .packages(), 'bookdown', 'knitr', 'rmarkdown'\n",
    "), 'packages.bib')"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "tags,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
