{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28667960",
   "metadata": {},
   "source": [
    "# Visualizing Genomes\n",
    "\n",
    "![Compressed de Bruijn graphs “hairballs”: https://academic.oup.com/bioinformatics/article/30/24/3476/2422268](./Figures/hairballsB.png){width=100%}\n",
    "\n",
    "## Set up Directories\n",
    "\n",
    "1. Make sure you're working in a **screen**\n",
    "\n",
    "2. Make directory\n",
    "\n",
    "```\n",
    "mkdir ~/viz\n",
    "```\n",
    "3. Navigate to the directory\n",
    "\n",
    "```\n",
    "cd ~/viz\n",
    "```\n",
    "4. Link to data\n",
    "\n",
    "```\n",
    "ln -s /home/data/pangenomics-2402/yprp/ .\n",
    "```\n",
    "\n",
    "## Graphical Fragment Assembly (GFA) format\n",
    "\n",
    "+ Originally developed for representing genomes during assembly\n",
    "+ Now used for pangenomics \n",
    "+ More on this later...\n",
    "\n",
    "## Bandage\n",
    "\n",
    "![Bandage: https://rrwick.github.io/Bandage/](./Figures/Bandage.png){width=100%}\n",
    "\n",
    "+ BLAST integration\n",
    "  + Can build a local BLAST database of the graph\n",
    "  + Can do a web BLAST search with sequences from nodes\n",
    "2. More details on making CSV labels: https://github.com/rrwick/Bandage/wiki/CSV-labels\n",
    "\n",
    "### **Group exercise:**{-}\n",
    "\n",
    "1. Copy the following example graph from inbre to your computer:\n",
    "```\n",
    "inbre.ncgr.org:/home/<username>/viz/yprp/example/S288C.SK1.minigraph.gfa\n",
    "```\n",
    "2. Open Bandage and load the graph\n",
    "3. Spend some time exploring the graph\n",
    "4. If your have BLAST, find [CUP1](https://www.yeastgenome.org/locus/S000001095) and [YHR054C](https://www.yeastgenome.org/locus/S000001096) via BLAST\n",
    "  + How many copies are in the graph?\n",
    "  + What does the structure it’s in look like?\n",
    "  + Take a screenshot of the region that CUP1 is in with the gene colored\n",
    "\n",
    "## General Feature Format (GFF)\n",
    "\n",
    "https://genome.ucsc.edu/FAQ/FAQformat.html#format3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49593a94",
   "metadata": {},
   "source": [
    "+ Plain text file\n",
    "+ 3 different versions\n",
    "+ Each line represents a feature\n",
    "+ 9 Columns\n",
    "  + Tab separated\n",
    "  + First 7 are the same for all feature types\n",
    "    + seqname, source, feature, start, end, score, strand\n",
    "  + 8th column is phase of CDS (coding DNA sequence) features\n",
    "    + 0, 1, or 2 for CDS features, . otherwise\n",
    "  + 9th column is for additional attributes related to feature\n",
    "\n",
    "## Browser Extensible Data (BED) Format\n",
    "\n",
    "https://genome.ucsc.edu/FAQ/FAQformat.html#format1\n",
    "\n",
    "+ Plain text file\n",
    "+ Designed for drawing features in genome browsers\n",
    "+ Each line represents a genomic region and associated annotations\n",
    "  + Features aren’t necessary biological\n",
    "+ 12 Columns\n",
    "  + Tab or white-space separated\n",
    "  + First 3 are required\n",
    "    + chrom, chromStart, chromEnd\n",
    "  + The next 9 are optional\n",
    "    + name, score, strand, thickStart, thickEnd, itemRgb, blockCount, blockSizes, blockStart\n",
    "\n",
    "## Integrative Genomics Viewer (IGV)\n",
    "\n",
    "https://software.broadinstitute.org/software/igv/\n",
    "\n",
    "+ View GFF/BED files relative to FASTA and multiple sequence alignments\n",
    "+ Doesn’t work with with names that contain dots “.”\n",
    "  + This conflicts with our naming convention, but we’ll work around it...\n",
    "\n",
    "### **Group exercise:**{-}\n",
    "\n",
    "1. Open IGV with S288C and GFF from [YPRP](https://yjx1217.github.io/Yeast_PacBio_2016/data/)\n",
    "2. Find [CUP1](https://www.yeastgenome.org/locus/S000001095) and [YHR054C](https://www.yeastgenome.org/locus/S000001096)\n",
    "    + How many copies does it have?\n",
    "\n",
    "## Linking IGV with Bandage\n",
    "\n",
    "1. Convert graph from GFA to BED\n",
    "\n",
    "```\n",
    "gfatools gfa2bed yprp/example/S288C.SK1.minigraph.gfa > S288C.SK1.minigraph.bed\n",
    "\n",
    "```\n",
    "\n",
    "### **Group Exercise**{-}\n",
    "\n",
    "1. Find an interesting structure in Bandage\n",
    "2. Get its node ID(s)\n",
    "3. `grep` the BED file for the ID(s)\n",
    "\n",
    "Why isn’t this BED file going to work in IGV?\n",
    "+ How should we solve this issue?\n",
    "\n",
    "4. Implement *your* solution\n",
    "5. Are there any other nuances you notice in the BED file?\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "-all",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
