{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fb116352",
   "metadata": {},
   "source": [
    "# Reference Graphs with VG\n",
    "\n",
    "## Variation Graph Toolkit (VG)\n",
    "\n",
    "+ Constructs graphs\n",
    "+ Manipulates graphs\n",
    "+ Indexes graphs\n",
    "+ Maps sequences to graphs\n",
    "+ Calls variants on mapped sequences\n",
    "+ Visualizes graphs\n",
    "\n",
    "![Garrison, E., Sirén, J., Novak, A. et al.](./Figures/VGref.png){width=65%}\n",
    "\n",
    "![vg Graph Genomics Pipeline: https://github.com/vgteam/vg](./Figures/VGpipe.png){width=100%}\n",
    "\n",
    "Also does:\n",
    "\n",
    "+ [Transcriptomic analysis](https://github.com/vgteam/vg#transcriptomic-analysis)\n",
    "+ Assembly-based pipelines\n",
    "+ So much more\n",
    "\n",
    "Graphs are cyclic, but otherwise general, i.e. reference, iterative, reference-free, etc.\n",
    "\n",
    "## Reference Graphs\n",
    "\n",
    "A reference genome \"decorated\" with variants:\n",
    "\n",
    "![GRAF™ Pan Genome Reference: https://www.sevenbridges.com/graf/](./Figures/GRAF.png){width=100%}\n",
    " \n",
    "## Pipeline\n",
    "\n",
    "1. Prepare input\n",
    "2. Construct graph\n",
    "3. View with Bandage\n",
    "\n",
    "## Preparing the Input\n",
    "\n",
    "Already **done** for you:\n",
    "\n",
    "+ Reference FASTA\n",
    "  + Description lines should only contain chromosome names\n",
    "  + Remove extraneous sequences, e.g. mitochondrial DNA\n",
    "+ VCF\n",
    "  + Ideally chromosome names exactly match chromosome names in reference\n",
    "\n",
    "## Set up directories\n",
    "\n",
    "1. Make sure you're working in a **screen**\n",
    "2. Make directory\n",
    "\n",
    "```\n",
    "mkdir ~/vg\n",
    "```\n",
    "3. Navigate to the directory\n",
    "\n",
    "```\n",
    "cd ~/vg\n",
    "```\n",
    "4. Link to data\n",
    "\n",
    "```\n",
    "ln -s /home/data/pangenomics-2402/yprp .\n",
    "```\n",
    "\n",
    "## Variant Call (VCF) Format\n",
    "\n",
    "https://en.wikipedia.org/wiki/Variant_Call_Format \n",
    "\n",
    "+ Plain text file\n",
    "+ Header:\n",
    "  + Header lines start with #\n",
    "  + Header lines with special keywords start with ##\n",
    "+ Columns\n",
    "  + Tab separated\n",
    "  + 8 mandatory columns\n",
    "    + CHROM, POS, ID, REF, ALT, QUAL , FILTER, INFO\n",
    "  + 9th column is optional\n",
    "    + FORMAT\n",
    "  + Each additional column represents a sample and has the format described in the FORMAT column\n",
    " \n",
    "## Variation Graph (VG) Format\n",
    "\n",
    "https://github.com/vgteam/vg/wiki/File-Formats\n",
    "\n",
    "+ [Protocol Buffer](https://protobuf.dev) messages (protobufs) describing a graph that can store sequence in its nodes and paths\n",
    "+ Encoded as binary\n",
    "+ Code for reading/writing protobufs can be generated for\n",
    "many languages\n",
    "+ [vg.proto](https://github.com/vgteam/libvgio/blob/master/deps/vg.proto)\n",
    "\n",
    "## Yeast Data\n",
    "\n",
    "Reference: **S288C**\n",
    "\n",
    "VCF: [3 SV calling pipelines on the 11 non-reference strains](https://link.springer.com/article/10.1186/s13059-020-1941-7)\n",
    "\n",
    "+ [Minimap2](https://github.com/lh3/minimap2) and paftools call\n",
    "+ [LAST](https://last.cbrc.jp) and [AsmVar](https://github.com/bioinformatics-centre/AsmVar)\n",
    "+ [nucmer](https://github.com/mummer4/mummer) and [Assemblytics](https://github.com/marianattestad/assemblytics)\n",
    "\n",
    "This workshop used an updated version of the Phase 1 pipeline available [here](https://github.com/alancleary/sv-genotyping-paper/tree/master/yeast#1-prepare-yeast-assemblies-and-call-svs-relative-to-the-reference-strain-s288c)\n",
    "\n",
    "## Construct Graph\n",
    "\n",
    "Construct a graph from the reference and VCF:\n",
    "```\n",
    "vg construct -r yprp/assemblies/S288C.genome.fa -v yprp/variants/S288C.vcf -S -a -f -p -t 20 > S288C.vg\n",
    "```\n",
    "\n",
    "+ **-r yprp/assemblies/S288C.fa**\n",
    "    + the reference genome\n",
    "+ **-v yprp/variants/S288C.sv.vcf**\n",
    "    + the VCF file containing variants (indels, structures, etc)\n",
    "+ **-S**\n",
    "    + include structural variants in construction of graph\n",
    "+ **-a**\n",
    "    + save paths for alts of variants by variant ID\n",
    "+ **-f**\n",
    "    + don't chop up alternate alleles from input VCF\n",
    "+ **-p**\n",
    "    + show progress\n",
    "+ **-t 20** \n",
    "    + use 20 threads\n",
    "+ **> S288C.vg**\n",
    "    + write the standard out (the computed graph) into S288C.vg\n",
    "\n",
    "Compact chains of nodes:\n",
    "```\n",
    "vg mod -u S288C.vg -t 20 > S288C.unchop.vg\n",
    "```\n",
    "\n",
    "+ **-u**\n",
    "    + unchop nodes, i.e. compact chains\n",
    "\n",
    "## Viewing with Bandage\n",
    "\n",
    "Convert the graph from vg to gfa:\n",
    "```\n",
    "vg view S288C.unchop.vg -g > S288C.unchop.gfa\n",
    "```\n",
    "\n",
    "+ **-g**\n",
    "    + output GFA format\n",
    "\n",
    "### Exercise\n",
    "\n",
    "1. Open with Bandage\n",
    "2. Where’s CUP1?\n",
    "3. Does the graph capture any CNV?\n",
    "4. What does it look like?\n",
    "5. Take a screenshot of the CUP1 region\n",
    " \n",
    "## VG Index Formats\n",
    "\n",
    "https://github.com/vgteam/vg/wiki/File-Formats\n",
    "\n",
    "XG (lightweight graph / path representation)\n",
    "\n",
    "+ Binary file containing graph structure (nodes, edges, paths) but no sequences\n",
    "+ Complex data structure that answers graph queries efficiently\n",
    "\n",
    "GCSA (Generalized Compressed Suffix Array)\n",
    "\n",
    "+ Equivalent to .sa file created by bwa index\n",
    "+ Binary file containing a suffix array that efficiently looks up where sequences occur in the graph\n",
    "\n",
    "## Indexing\n",
    "\n",
    "Generate .xg and .gcsa files (1min):\n",
    "```\n",
    "vg index -x S288C.xg -g S288C.gcsa S288C.vg -t 20\n",
    "```\n",
    "\n",
    "+ **-x S288C.xg**\n",
    "    + the output .xg file\n",
    "+ **-g S288C.gcsa** \n",
    "    + the output GCSA file\n",
    "+ **-t 20** \n",
    "    + use 20 threads\n",
    "    \n",
    "NOTE: **Don’t** index the **unchopped** file!\n",
    "\n",
    "NOTE: You can index a GFA file but this may have implications for mapping.\n",
    "There’s also an [autoindex](https://github.com/vgteam/vg/wiki/Automatic-indexing-for-read-mapping-and-downstream-inference) command.\n",
    "\n",
    "## Graph Alignment/Map (GAM) Format\n",
    " \n",
    "https://github.com/vgteam/vg/wiki/File-Formats\n",
    "\n",
    "+ Analogous to BAM, but for graphs\n",
    "+ Binary file describing where reads mapped to in the graph\n",
    "structure\n",
    "+ Uncompressed has one read per line\n",
    "+ Can be converted to JSON for manual parsing (very inefficient!)\n",
    "\n",
    "## Read Mapping\n",
    "\n",
    "Map reads (9min):\n",
    "```\n",
    "vg map -d S288C -f yprp/reads/SK1.illumina.fastq.gz -t 20 > S288C.SK1.illumina.gam\n",
    "```\n",
    "\n",
    "+ **-d S288C**\n",
    "    + use the S288C.xg and S288C.gcsa files\n",
    "+ **-f yprp/reads/SK1.illumina.fastq.gz**\n",
    "    + the reads file\n",
    "+ **-t 20**\n",
    "    + use 20 threads\n",
    "+ **S288C.SK1.illumina.gam**\n",
    "    + the output GAM file\n",
    "\n",
    "Computing mapping stats (<1min):\n",
    "\n",
    "```\n",
    "vg stats -a S288C.SK1.illumina.gam\n",
    "```\n",
    "\n",
    "+ **-a**\n",
    "    + input is an alignment (GAM) file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cea3eb6e",
   "metadata": {},
   "source": [
    "##  Bringing Alignments Back to Single Genomes\n",
    "\n",
    "Map reads (1min):\n",
    "\n",
    "```\n",
    "vg surject -x S288C.xg -b S288C.SK1.illumina.gam -t 20 > S288C.SK1.illumina.BAM\n",
    "```\n",
    "+ **-x S288C.xg**\n",
    "    + graph or xg to use\n",
    "+ **-b**\n",
    "    + BAM output\n",
    "+ **-t 20**\n",
    "    + use 20 threads\n",
    "\n",
    "### **Exercise:**{-}\n",
    "\n",
    "1. Rename chromosomes in BED for use with IGV\n",
    "2. Copy BED to personal computer\n",
    "3. Spend some time looking at BED in IGV and share any interesting observations\n",
    "you may have? \n",
    "+ What’s up with CUP1?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c09468a",
   "metadata": {},
   "source": [
    "##  Preparing the BAM IGV (or other genome viewer)\n",
    "\n",
    "**Only necessary if genome viewer can’t match UCSC style names names in BAM to chromosomes in reference.**\n",
    "\n",
    "1. Convert BAM to SAM (<1min):\n",
    "\n",
    "```\n",
    "samtools view -h -o S288C.SK1.illumina.sam S288C.SK1.illumina.BAM\n",
    "```\n",
    "\n",
    "2. Remove assembly name (<1min):\n",
    "\n",
    "```\n",
    "sed 's/S288C.//' S288C.SK1.illumina.sam > S288C.SK1.illumina.renamed.sam\n",
    "```\n",
    " \n",
    "3. Convert SAM to BAM (6min):\n",
    "\n",
    "```\n",
    "samtools view -bS S288C.SK1.illumina.renamed.sam > S288C.SK1.illumina.renamed.bam\n",
    "```\n",
    "<===\n",
    " \n",
    "4. Sort the bam file (6min):\n",
    "\n",
    "```\n",
    "samtools sort -o S288C.SK1.illumina.renamed.sorted.bam S288C.SK1.illumina.renamed.bam\n",
    "```\n",
    "\n",
    "5. Index BAM (<1min):\n",
    "\n",
    "```\n",
    "samtools index S288C.SK1.illumina.renamed.sorted.bam\n",
    "```\n",
    "\n",
    "## Pack (pileup support) Format\n",
    "\n",
    "https://github.com/vgteam/vg/wiki/File-Formats\n",
    "\n",
    "+ Binary file\n",
    "+ Computes pileup support for each read in a mapping\n",
    "+ The format isn’t actually documented...\n",
    "\n",
    "## Calling Graph Supported Variants\n",
    "\n",
    "Compute read support for variation already in the graph (<1min):\n",
    "\n",
    "```\n",
    "vg pack -x S288C.xg -g S288C.SK1.illumina.gam -Q 5 -s 5 -o S288C.SK1.illumina.pack -t 20\n",
    "```\n",
    "\n",
    "+ **-Q 5**\n",
    "    + ignore mapping and base quality < 5\n",
    "+ **-s 5**\n",
    "    + ignore the first and last 5pb of each read\n",
    "+ **-o S288C.SK1.illumina.pack**\n",
    "    + the output pack file\n",
    "+ **-t 20**\n",
    "    + use 20 threads\n",
    "    \n",
    "Generate a VCF from the support (<1min):\n",
    "\n",
    "```\n",
    "vg call S288C.xg -k S288C.SK1.illumina.pack -t 20 > S288C.SK1.illumina.graph_calls.vcf\n",
    "```\n",
    "\n",
    "## Calling Novel Variants\n",
    "\n",
    "Augment the graph with the mapped reads (45min):\n",
    "\n",
    "```\n",
    "vg augment S288C.vg S288C.SK1.illumina.gam -A S288C.SK1.illumina.aug.gam -t 20 > S288C.SK1.illumina.aug.vg\n",
    "```\n",
    "\n",
    "+ *NOTE: Only supports VG files and indexes used for mapping must be built from the same VG file being augmented, i.e. indexes build from GFA files that were then converted to VG won’t work.*\n",
    "\n",
    "Index the augmented graph (1min):\n",
    "\n",
    "``` \n",
    "vg index -x S288C.SK1.illumina.aug.xg S288C.SK1.illumina.aug.vg -t 20\n",
    "```\n",
    "\n",
    "Compute read support for novel variation (2min):\n",
    "\n",
    "```\n",
    "vg pack -x S288C.SK1.illumina.aug.xg -g S288C.SK1.illumina.aug.gam -Q 5 -s 5 -o S288C.SK1.illumina.aug.pack -t 20\n",
    "```\n",
    "\n",
    "Generate a VCF from the support (3min)\n",
    "\n",
    "```\n",
    "vg call S288C.SK1.illumina.aug.xg -k S288C.SK1.illumina.aug.pack -t 20 > S288C.SK1.illumina.aug_calls.vcf\n",
    " ```\n",
    " \n",
    "## Calling Variants Already in the Graph\n",
    "\n",
    "Output variants used to construct graph (<1min):\n",
    "\n",
    "```\n",
    "vg deconstruct S288C.xg -P S288C -t 20 > S288C.deconstruct.vcf\n",
    " ```\n",
    " \n",
    "+ **-P S288C**\n",
    "    + report variants relative to paths with names that start with S288C\n",
    "\n",
    "**`S288C.deconstruct.vcf` might not be identical to S288C.vcf because VG takes liberties with variants when constructing the graph.**\n",
    "\n",
    "##  Pros and Cons Reference Graphs\n",
    "\n",
    "**Pros:**\n",
    "\n",
    "+ Graphs are relatively small\n",
    "  + Easy to inspect visually\n",
    "  + Good for low-quality / fragmented assemblies\n",
    "  + Shorter run-times\n",
    "\n",
    "**Cons:**\n",
    "\n",
    "+ Variation represented is determined by VCF pipeline\n",
    "  + Only describes variation relative to the reference\n",
    "  + Suboptimal mappings\n",
    "  + Suboptimal population structure\n",
    " \n",
    "##  Additional vg Commands that are Useful\n",
    "\n",
    "### vg combine\n",
    "\n",
    "Combines two or more graphs into one.\n",
    "\n",
    "+ Use to construct graph in parts (e.g. individual chromosomes) and then combine\n",
    "+ Use to combine graphs of related species\n",
    "\n",
    "### Modifying and Simplifying Graphs\n",
    "\n",
    "**vg chunk** - Splits graphs or alignments into smaller chunks. \n",
    "\n",
    "**vg mod** - Used to modify graphs.\n",
    "\n",
    "  + We’ve already used vg mod -u to “unchop” nodes. \n",
    "\n",
    "**vg simplifying** - Simplifies a graph.\n",
    "\n",
    "**vg find** - Used to identify subgraphs based on various metrics and parameters.\n",
    "\n",
    "  + e.g. anchored by a region on a specific chromosome. \n",
    "\n",
    "Use **--help** with any command to learn how to use it.\n",
    "\n",
    "## Drawing Graphs \n",
    "\n",
    "**vg view** - Outputs graphs and other structures for drawing.\n",
    "\n",
    "  + We’ve already used vg *view -g* to convert from vg to gfa for viewing in Bandage.\n",
    "  + Can also output dot (among other files types) which is a more general graph drawing format (compatible with [Graphviz](https://graphviz.org)).\n",
    "\n",
    "**vg viz** - Basic graph drawing ([documentation](https://github.com/vgteam/vg/wiki/Visualization)).\n",
    "  + Recommend only using on **small (sub)graphs**.\n",
    "  + Use in conjunction with *chunk*, *mod*, and/or *find* to draw an interesting subgraph.\n",
    "  + You can include mapped reads in the visualization!\n",
    " \n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
