{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2f4aade1",
   "metadata": {},
   "source": [
    "# Reference-Free Graphs with Cactus\n",
    "\n",
    "## Cactus\n",
    "\n",
    "https://github.com/ComparativeGenomicsToolkit/cactus\n",
    "\n",
    "+ Reference-free whole genome MSA \n",
    "+ Constructs graph based on MSA\n",
    " \n",
    "## Cactus Graphs\n",
    "\n",
    "Cactus Graphs “naturally decompose the common substructures in a set of related genomes into a hierarchy of chains that can be visualized as two-dimensional multiple alignments and nets that can be visualized in circular genome plots”\n",
    "\n",
    "https://www.liebertpub.com/ doi/abs/10.1089/cmb.2010.02 52\n",
    "\n",
    "![Cactus](./Figures/Cactus.png){width=100%} \n",
    "\n",
    "##  Cactus Algorithm\n",
    "\n",
    "1. Multiple sequence aligner\n",
    "2. Originally developed for multi-species alignments\n",
    "3. Fast because it uses a guide tree (Newick format)\n",
    "+ https://evolution.genetics.washington.edu/phylip/newicktree.html\n",
    "4. Now supports minigraph GFA in place of guide tree for pangenome\n",
    "alignments\n",
    "+ https://github.com/lh3/minigraph\n",
    "\n",
    "![Yeast](./Figures/Yeast.png){width=100%}\n",
    "\n",
    "## Reference-Free Graphs\n",
    "\n",
    "https://academic.oup.com/bioinformatics/article/30/24/3476/2422268\n",
    "\n",
    "![Input Genomes](./Figures/InputGenomes.png){width=100%}\n",
    "\n",
    "## Pipeline\n",
    "\n",
    "1. minigraph\n",
    "2. Prepare the input\n",
    "3. cactus\n",
    "4. vg\n",
    "5. View with Bandage\n",
    "\n",
    "### Set up Directories\n",
    "\n",
    "1. Make sure you're working in a **screen**\n",
    "\n",
    "2. Make sure you've sourced the pangenomics environment file\n",
    "```\n",
    "source /home/pangenomics/pangenomics_env\n",
    "```\n",
    "\n",
    "3. Make Directory\n",
    "\n",
    "```\n",
    "mkdir cactus\n",
    "```\n",
    "4. Navigate to the Directory\n",
    "\n",
    "```\n",
    "cd cactus\n",
    "```\n",
    "5. Link to data\n",
    "\n",
    "```\n",
    "cp -r /home/pangenomics/data/yprp/assemblies .\n",
    "```\n",
    "\n",
    "*Note:* Don't use \"ln -s /home/pangenomics/data/yprp/ .\"\n",
    "\n",
    "## Yeast Data\n",
    "\n",
    "Reference:\n",
    "+ S288C\n",
    "\n",
    "Using all 12 YPRP assemblies\n",
    "\n",
    "## Preparing the Input\n",
    "\n",
    "**(already done for you)**\n",
    "\n",
    "1. FASTA files\n",
    "+ Chromosome names should be unique across files\n",
    "    + We’re using: \n",
    "    \n",
    "    ```\n",
    "    <strain name>.<chromosome>*\n",
    "    ```\n",
    "    \n",
    "E.G.\n",
    "\n",
    ">S288C.chrI\n",
    "\n",
    "## minigraph\n",
    "\n",
    "Use the graph we previously built: yprp.minigraph.gfa\n",
    "\n",
    "## Preparing the Input (exercise)\n",
    "\n",
    "1. Cactus seqFile tells Cactus where to load sequences from\n",
    "+ Maps sequence names to file paths\n",
    "    + We’re using: \n",
    "    \n",
    "    ```\n",
    "    “strain name>\\t<path to sequence>”\n",
    "    ```\n",
    "\n",
    "    + Must include “_MINIGRAPH_” entry for path to minigraph GFA \n",
    "\n",
    "E.G.\n",
    "\n",
    "seqFile:\n",
    "S288C ./S288C.genome.fa _MINIGRAPH_ yprp.minigraph.gfa\n",
    "\n",
    "+ It’s recommended the minigraph contains all the sequences in the seqFile\n",
    "\n",
    "2. Call it **yprp.seqFile.txt**\n",
    "\n",
    "## Cactus\n",
    "\n",
    "1. Align each input FASTA to the minigraph (2min):\n",
    "\n",
    "```\n",
    " cactus-graphmap jobStore yprp.seqFile.txt yprp.minigraph.gfa\n",
    " yprp.cactus.paf --outputFasta yprp.minigraph.gfa.fa\n",
    " --maxCores 20\n",
    " ```\n",
    "\n",
    "+ **jobstore**\n",
    "    + a directory where intermediate files should be stored (shouldn’t exist)\n",
    "+ **yprp.seqfile.txt**\n",
    "    + text file mapping sequence names to file paths\n",
    "+ **yprp.minigraph.gfa**\n",
    "    + the graph constructed by minigraph\n",
    "+ **yprp.cactus.paf**\n",
    "    + what to name the output pairwise mapping file\n",
    "+ **yprp.minigraph.gfa.fa**\n",
    "    + a FASTA to output the GFA’s sequence to \n",
    "*NOTE:* This command modifiers the seqFile. Make a copy before running!\n",
    "\n",
    "2. Generate multiple alignment and VG graph (19min):\n",
    "\n",
    "*NOTE*: Don’t run this command if you intend to use\n",
    "\n",
    "```\n",
    " cactus-graphmap-split\n",
    " cactus-align jobStore yprp.seqFile.txt yprp.cactus.paf\n",
    " yprp.cactus.hal --pangenome --pafInput --outVG\n",
    " --reference S288C --maxCores 20\n",
    " ```\n",
    "\n",
    "+ **yprp.cactus.hal**\n",
    "    + what to name the output multiple sequence alignment file\n",
    "+ **reference S288C**\n",
    "    + the name of the reference in the seqFile (should be the same as\n",
    "the reference in the minigraph GFA)\n",
    "\n",
    "## Preparing the Input (exercise)\n",
    "\n",
    "1. Cactus contigs files tells Cactus to contigs to split the graph on\n",
    "+ A list of all the S288C contigs\n",
    "+ Call it S288C.contigs.txt\n",
    "\n",
    "### Preparing the Input (solution)\n",
    "\n",
    "Make a reference contigs file:\n",
    "\n",
    "```\n",
    "grep -Po \"^>\\K.*\" ~/cactus/yprp/assemblies/S288C.genome.fa >\n",
    "S288C.contigs.txt\n",
    "```\n",
    "\n",
    "## Cactus Split\n",
    "\n",
    "1. Align each input FASTA to the minigraph and split by reference chromosome (4min):\n",
    "\n",
    "```\n",
    " cactus-graphmap-split jobStore yprp.seqFile.txt\n",
    " yprp.minigraph.gfa yprp.cactus.paf\n",
    " --refContigsFile S288C.contigs.txt --reference S288C\n",
    " --outDir chroms --maxCores 20\n",
    " ```\n",
    " \n",
    "+ **yprp.seqFile.txt**\n",
    "    + modified version from previous *cactus-graphmap* command\n",
    "+ **yprp.cactus.paf**\n",
    "    + output from previous *cactus-graphmap* command\n",
    "+ **-refContigsFile**\n",
    "    + the names of the chromosomes in the reference FASTA\n",
    "+ **–reference**\n",
    "    + the name of the reference in the seqFile\n",
    "+ **-outDir**\n",
    "    + where the split outputs should be placed\n",
    "\n",
    "2. Generate multiple alignment and graph for a chromosome (35min total):\n",
    "\n",
    "```\n",
    "cactus-align jobStore chroms/seqfiles/S288C.chrI.seqFile\n",
    "chroms/S288C.chrI/S288C.chrI.paf\n",
    "yprp.S288C.chrI.cactus.hal --pangenome --pafInput --outVG\n",
    "--reference S288C --maxCores 20\n",
    "```\n",
    "\n",
    "+ Recommend automating with a bash script; \n",
    "    + see: /home/pangenomics/scripts/cactus-align-chromosomes.sh\n",
    "\n",
    "*NOTE*: Other options are:\n",
    "\n",
    "+ cactus-align --batch\n",
    "+ cactus-align-batch\n",
    "\n",
    "## Viewing with Bandage\n",
    "\n",
    "View **one chromosome at a time** with Bandage\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
