{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a681dfcf",
   "metadata": {},
   "source": [
    "# Other Tools and Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "525d587f",
   "metadata": {},
   "source": [
    "## Workshop Tools Review\n",
    "\n",
    "+ **Bandage**\n",
    "    + https://rrwick.github.io/Bandage/\n",
    "    + https://academic.oup.com/bioinformatics/article/31/20/3350/196114\n",
    "+ **IGV**\n",
    "    + https://software.broadinstitute.org/software/igv/\n",
    "    + https://www.nature.com/articles/nbt.1754\n",
    "+ **vg**\n",
    "    + https://github.com/vgteam/vg\n",
    "    + https://www.nature.com/articles/nbt.4227\n",
    "+ **minigraph**\n",
    "    + https://github.com/lh3/minigraph\n",
    "    + https://link.springer.com/article/10.1186/s13059-020-02168-z\n",
    "+ **cactus**\n",
    "    + https://github.com/ComparativeGenomicsToolkit/cactus\n",
    "    + https://www.nature.com/articles/s41586-020-2871-y\n",
    "+ **pggb**\n",
    "    + https://github.com/pangenome/pggb\n",
    "    + https://www.biorxiv.org/content/1\n",
    "+ **cuttlefish**\n",
    "    + https://github.com/COMBINE-lab/cuttlefish\n",
    "    + https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02743-6\n",
    "+ **gfatools**\n",
    "    + https://github.com/lh3/gfatools\n",
    "+ **samtools**\n",
    "    + http://www.htslib.org/\n",
    "\n",
    "## Other Visualizations\n",
    "\n",
    "### Bandage with Paths\n",
    "\n",
    "https://github.com/asl/Bandage\n",
    "\n",
    "A fork of Bandage that supports embedded paths\n",
    "\n",
    "+ graphs constructed with minigraph **don’t have paths**!\n",
    "    \n",
    "![Fork of a Bandage](./Figures/Fork.png){width=100%}\n",
    "\n",
    "### GfaViz\n",
    "\n",
    "https://github.com/ggonnella/gfaviz \n",
    "\n",
    "+ Manuscript:\n",
    "    + https://academic.oup.com/bioinformatics/article/35/16/2853/5267826?login=false\n",
    "\n",
    "A GFA viewer that explicitly supports variation graphs (Fig c).\n",
    "\n",
    "![Example Applications of GfaViz](./Figures/Fig2.png){width=100%}\n",
    "\n",
    "### Panache \n",
    "\n",
    "https://github.com/SouthGreenPlatform/panache\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://academic.oup.com/bioinformatics/article/37/23/4556/6380547\n",
    "\n",
    "Web browser-based visualization of linearized pangenomes\n",
    "\n",
    "![Panache Genome Representations](./Figures/Panache.png){width=100%}\n",
    "\n",
    "### Anvi'o\n",
    "\n",
    "https://anvio.org\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://peerj.com/articles/1319/\n",
    "+ Pangenomics Workflow:\n",
    "    + https://merenlab.org/2016/11/08/pangenomics-v2/\n",
    "\n",
    "An analysis and visualization platform that can run pipelines on microbial pangenomic data.\n",
    "\n",
    "![Anvi'o Genome Bins for infant gut ](./Figures/fig-2-full.png){width=100%}\n",
    "\n",
    "![Anvi'o](./Figures/Anvio.png){width=100%}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c3bbf7c",
   "metadata": {},
   "source": [
    "### Visualizing Alignment of Microbial Genomes\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://peerj.com/articles/9576/\n",
    "    \n",
    "Sequence-level alignment showing common and rare variations at sequence level.\n",
    "\n",
    "![Seq-level Alignment](./Figures/MicrobialSeqAlign.png){width=80%}\n",
    "\n",
    "### PanX\n",
    "\n",
    "https://github.com/neherlab/pan-genome-analysis\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://academic.oup.com/nar/article/46/1/e5/4564799\n",
    "+ Demo:\n",
    "    + https://pangenome.org\n",
    "    \n",
    "Pangenome analysis pipeline and visualizations.\n",
    "\n",
    "![Interconnected Components of PanX](./Figures/PanXFig.png){width=100%}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "190a5227",
   "metadata": {},
   "source": [
    "### Sequence Tube Maps\n",
    "\n",
    "https://github.com/vgteam/sequenceTubeMap\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://academic.oup.com/bioinformatics/article/35/24/5318/5542397\n",
    "+ Demo:\n",
    "    + https://vgteam.github.io/sequenceTubeMap/\n",
    "\n",
    "![Seq Tube](./Figures/SeqTube.png){width=100%}\n",
    "\n",
    "### MoMI-G\n",
    "\n",
    "https://github.com/MoMI-G/MoMI-G\n",
    "\n",
    "+ Manuscript\n",
    "    + https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-3145-2\n",
    "+ Demo:\n",
    "    + http://demo.momig.tokyo\n",
    "\n",
    "Web-based multi-scale genome graph browser.\n",
    "\n",
    "![MoMI-G](./Figures/momig1.png){width=100%}\n",
    "\n",
    "![MoMI-G](./Figures/momig2.png){width=50%}\n",
    "\n",
    "### PGV (Pangenome Graph Viewer)\n",
    "\n",
    "https://github.com/ucrbioinfo/PGV\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04424-w\n",
    "+ Demo:\n",
    "    + https://panviz-22b36.firebaseapp.com \n",
    "\n",
    "Web-based reference-agnostic representation and visualization of pan-genomes constructed using a **progressiveMauve** pipeline.\n",
    "\n",
    "+ https://darlinglab.org/mauve/user-guide/progressivemauve.html\n",
    "\n",
    "![PGV](./Figures/PGV.png){width=100%}\n",
    "\n",
    "![PGV](./Figures/PGV2.png){width=100%}\n",
    "\n",
    "### panGraphViewer\n",
    "\n",
    "https://github.com/TF-Chan-Lab/panGraphViewer\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://www.biorxiv.org/content/biorxiv/early/2023/04/02/2023.03.30.534931.full.pdf\n",
    "    \n",
    "Pangenome graph visualizer that can incorporate annotation.\n",
    "\n",
    "![panGraphViewer](./Figures/panGraphViewer.jpeg){width=80%}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d82da39",
   "metadata": {},
   "source": [
    "## Other Graph Construction / Manipulation Tools / Pipelines\n",
    "\n",
    "### odgi (optimized dynamic genome/graph implementation)\n",
    "\n",
    "https://github.com/pangenome/odgi\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://www.biorxiv.org/content/10.1101/2021.11.10.467921v1\n",
    "+ Documentation:\n",
    "    + https://odgi.readthedocs.io/en/latest/\n",
    "    \n",
    "“A toolkit for understanding pangenome graphs.”\n",
    "\n",
    "Functionality:\n",
    "\n",
    "1. Explore graphs via 1D and 2D visualization\n",
    "2. Detect complex regions in the graph\n",
    "3. Extract selected loci\n",
    "4. Sorting and laying out, e.g. in preparation for visualization\n",
    "5. Navigating and annotating graphs, e.g. bringing annotations in\n",
    "to Bandage\n",
    "6. Remove artifacts and complex regions\n",
    "7. MultiQC report for graph statistics\n",
    "\n",
    "### seqwish\n",
    "\n",
    "https://github.com/ekg/seqwish\n",
    "\n",
    "+ Dissertation\n",
    "    + https://github.com/ekg/seqwish\n",
    "    \n",
    "“The sequences wish they were squished into a graph.” \n",
    "\n",
    "+ Converts pairwise alignments between sequences into a variation graph, i.e. pangenome.\n",
    "\n",
    "+ Takes one or more PAF files as input (e.g. output of minimap2)\n",
    "+ Infers graph nodes and node orderings from alignment blocks\n",
    "and how they overlap between sequences.\n",
    "\n",
    "###  smoothxg\n",
    "\n",
    "https://github.com/pangenome/smoothxg\n",
    "\n",
    "“Linearize and simplify variation graphs using blocked partial order alignment.”\n",
    "\n",
    "+ “Smooths” graph by consolidating structures that are artifacts of graph assembly / data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5d7efac",
   "metadata": {},
   "source": [
    "## Other Mapping, Variant Calling, and Genotyping Tools\n",
    "\n",
    "### GraphAligner\n",
    "\n",
    "https://github.com/maickrau/GraphAligner\n",
    "\n",
    "+ Manuscript\n",
    "    + https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02157-2\n",
    "    \n",
    "“Aligns long reads to genome graphs. Compared to the state-of-the-art tools, GraphAligner is 13x faster and uses 3x less memory. When employing GraphAligner for error correction, we find it to be more than twice as accurate and over 12x faster than extant tools.”\n",
    "\n",
    "### GraphTyper / GraphTyper2\n",
    "\n",
    "https://github.com/DecodeGenetics/graphtyper\n",
    "\n",
    "+ Manuscripts\n",
    "    + https://www.nature.com/articles/ng.3964\n",
    "    + https://www.nature.com/articles/s41467-019-13341-9\n",
    "\n",
    "“A graph-based variant caller capable of genotyping population-scale short read data sets. It represents a reference genome and known variants of a genomic region using an acyclic graph structure (a \"pangenome reference\"), which high-throughput sequence reads are re-aligned to for the purpose of discovering and genotyping SNPs, small indels, and structural variants.”\n",
    "\n",
    "+ GraphTyper2 is the v2 release of GraphTyper, i.e. the same GitHub repository\n",
    "\n",
    "### BayesTyper\n",
    "\n",
    "https://github.com/bioinformatics-centre/BayesTyper\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://www.nature.com/articles/s41588-018-0145-5\n",
    "\n",
    "Performs genotyping of all types of variation by exactly aligning sequence k-mers to a reference graph and then using Bayesian inference built on a “variation-prior” database to do genotyping.\n",
    "\n",
    "BayesTyper enables “sensitive and accurate genotyping across classes and varying lengths of structural variants” as compared to other methods\n",
    "\n",
    "### Paragraph\n",
    "\n",
    "https://github.com/Illumina/paragraph\n",
    "\n",
    "+ Manuscript\n",
    "    + https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1909-7\n",
    "\n",
    "A suite of graph-based genotyping tools to support accurate genotyping of short-read data.\n",
    "\n",
    "Developed by Illumina.\n",
    "\n",
    "“Paragraph has better accuracy than other existing genotypers and can be applied to population-scale studies.”\n",
    "\n",
    "### Practical Haplotype Graph (PHG)\n",
    "\n",
    "https://bitbucket.org/bucklerlab/practicalhaplotypegraph/wiki/Home\n",
    "\n",
    "+ Manuscript\n",
    "    + https://academic.oup.com/g3journal/article/12/2/jkab390/6423995\n",
    "    \n",
    "A graphical genotyping data storage and genotype imputation platform designed to be used on skim sequencing methods to infer high-density genotypes directly from low-coverage sequence.\n",
    "\n",
    "+ Developed by the Buckler Lab\n",
    "+ Has been used extensively with wheat and maize    \n",
    "\n",
    "### PanGenie\n",
    "\n",
    "https://github.com/eblerjana/pangenie\n",
    "\n",
    "+ Manuscript\n",
    "    + https://doi.org/10.1038/s41588-022-01043-w\n",
    "    \n",
    "A short-read genotyper for single nucleotide polymorphisms, indels, and structural variants.\n",
    "\n",
    "\"Genome inference\" using kmers from short reads plus haplotypes in the pangenome reference\n",
    "\n",
    "\"Compared with mapping-based approaches, PanGenie is more than 4 times faster at 30-fold coverage and achieves better genotype concordances for almost all variant types and coverages tested.\"\n",
    "\n",
    "## Gene Resolution Tools\n",
    "\n",
    "### Genome Context Viewer\n",
    "\n",
    "https://github.com/legumeinfo/gcv\n",
    "\n",
    "+ Manuscript:\n",
    "    + https://academic.oup.com/nar/article/46/1/e5/4564799\n",
    "+ Demo:\n",
    "  + https://www.legumeinfo.org/gcv2/instructions\n",
    "  \n",
    "Visual exploration of population using synteny derived from gene homology relationships (e.g. gene families)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adee91ea",
   "metadata": {},
   "source": [
    "![Viewer](./Figures/PanXFig2.png){width=100%}\n",
    "\n",
    "### pandagma\n",
    "\n",
    "https://github.com/legumeinfo/pandagma\n",
    "\n",
    "PANgenome Directed Acyclic Graph Multiple Alignment\n",
    "\n",
    "\"Pandagma is a collection of tools for calculating pangene sets and gene families.\"\n",
    "\n",
    "+ `pandagma pan` for pangene sets\n",
    "+ `pandagma fam` for gene families"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "154a0fb8",
   "metadata": {},
   "source": [
    "![Viewer](./Figures/PanXFig2.png){width=100%}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fbdb66e",
   "metadata": {},
   "source": [
    "## So Much More...\n",
    "\n",
    "New tools are coming out constantly!\n",
    "\n",
    "Check the citations of tools of interest to stay up-to-date!\n",
    "\n",
    "## Future Tools\n",
    "\n",
    "+ Data compression!\n",
    "+ This research is happening now\n",
    "  + Developing techniques that allow computation directly on compressed data\n",
    "  + Techniques specifically for pangenomes\n",
    "  + Fundamental data compression algorithms still need to be solver\n",
    "  + Usable software is still a few years away...\n",
    "\n",
    "## The Pangenome book, edited by Tettlin!\n",
    "\n",
    "https://library.oapen.org/handle/20.500.12657/37707"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
