{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pangenomics\n",
    "--------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intro to Graphical Pangenomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Overview\n",
    "\n",
    "Pangenome graphs are representations of related genomes that enable exploration of the relationships of the genomes to one another, their commonalities and novelties, and their collective genetic variation. You will learn about different types of pangenomic graphs and their strengths and weaknesses.\n",
    "\n",
    "A brief overview of pangenomes and this module is available in this video XXX."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Learning Objectives\n",
    "+ Describe pangenomic data\n",
    "+ Describe pangenomic graphs\n",
    "+ Specify the different types of pangenomic graphs and identify their pros and cons\n",
    "+ Detail the pipeline that will be used in this module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Started\n",
    "\n",
    "In this submodule you will learn about pangenomic data and different types of graphs, including their pros and cons. In addition, you will learn about the pangenomics pipeline that will be used in this module.\n",
    "\n",
    "#### Pangenomics pipeline\n",
    "- Build graphs\n",
    "- Map reads\n",
    "- Call variants\n",
    "- Visualize graphs and mapped reads3\n",
    "\n",
    "----------------\n",
    "\n",
    "## What is a \"pangenome\"?\n",
    "\n",
    "The term “pangenome” was first coined by [Sigaux et al. (2000)](https://pubmed.ncbi.nlm.nih.gov/11261250/) and was used to describe a public database containing an assessment of genome and transcriptome alterations in major types of tumors, tissues, and experimental models. The figure below shows the use of the term \"pan genome\" in the abstract of the [Sigaux et al. (2000)](https://pubmed.ncbi.nlm.nih.gov/11261250/) paper.\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/Abstract1.png\"\n",
    "    alt=\"Sigaux et al. abstract\" />\n",
    "  <figcaption><a href=\"https://pubmed.ncbi.nlm.nih.gov/11261250/\">https://pubmed.ncbi.nlm.nih.gov/11261250/</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "The term was later revitalized by [Tettelin et al. (2005)](https://pubmed.ncbi.nlm.nih.gov/16172379/) to describe a microbial genome by which genes were in the core (present in all strains) and which genes were dispensable (missing from one or more of the strains). The figure below shows the use of the term \"pan-genome\" in the abstract of the [Tettelin et al. (2005)](https://pubmed.ncbi.nlm.nih.gov/16172379/) paper. This paper also introduces the concept of a \"core genome\", consisting of genomic regions or genes that are present in all the strains that were analyzed and a \"dispensible genome\" consisting of regions or genes present in only one or some of the strains. \"Core genome\" sequences are presumed to be critical to the species because they have been retained by all of the strains.\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/Abstract2.png\"\n",
    "    alt=\"Tettelin et al. abstract\" />\n",
    "  <figcaption><a href=\"https://pubmed.ncbi.nlm.nih.gov/16172379/\">https://pubmed.ncbi.nlm.nih.gov/16172379/</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "\n",
    "While the terms \"core\" and \"dispensible\" or \"accessory\" genomes are still used, some people partition the pangenome into additional sections. For instance, the pangenome figure below ([EVCB Mx, 2021a](https://en.wikipedia.org/wiki/Pan-genome)) shows a pangenome partitioned into core (in all genomes), shell (in more than one but not all genomes), and cloud (in only a single accession) designations.\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/Pangenome.png\"\n",
    "    alt=\"Pangenome as Venn diagram\" />\n",
    "  <figcaption><a href=\"https://en.wikipedia.org/wiki/Pan-genome\">https://en.wikipedia.org/wiki/Pan-genome</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "### Open vs. Closed Genomes\n",
    "\n",
    "Pangenomes can be designated as open or closed. As shown in [EVCB Mx (2021a)](https://en.wikipedia.org/wiki/Pan-genome), closed pangenomes have a relatively large ratio of core to accessory regions. In these pangenomes, a small number of genomes is enough to capture all or nearly all of the genes or genomic regions present across the species. On the other hand, open pangenomes have a small ratio of core to accessory regions. As more genomes are sequenced and added to the pangenome, additional genes or genomic regions are found that had not been previously seen.\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/ClosedvOpen.png\"\n",
    "    alt=\"Open and Closed Pangenomes\" />\n",
    "  <figcaption><a href=\"https://en.wikipedia.org/wiki/Pan-genome\">https://en.wikipedia.org/wiki/Pan-genome</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "### Then vs. Now\n",
    "\n",
    "Pangenomes are becoming increasingly feasible because sequencing costs have dropped precipitously while throughput has increased rapidly. The figure below ([Wetterstrand, 2023](https://www.genome.gov/about-genomics/fact-sheets/DNA-Sequencing-Costs-Data)) shows the cost of sequencing a human genome tracking Moore's law until the Next Generation Sequencing Revolution, in which several new sequencing technologies were introduced, manifested around 2007/2008. Since then, the cost of sequencing a human genome has dropped approximately 10,000 fold.\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/CostGenome.png\"\n",
    "    alt=\"Cost per Genome\" />\n",
    "  <figcaption><a href=\"https://www.genome.gov/about-genomics/fact-sheets/DNA-Sequencing-Costs-Data\">https://www.genome.gov/about-genomics/fact-sheets/DNA-Sequencing-Costs-Data</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "\n",
    "These cost reductions, along with improvements in the generation of high quality long sequencing reads such as [PacBio Hifi](https://www.pacb.com/technology/hifi-sequencing/) reads, have enabled the sequencing of multiple, high quality genomes within many different species, faciliting the construction of pangenomes and resulting in a steady increase in the number of publications mentioning pangenomes [(Bayer et al., 2020)](https://www.nature.com/articles/s41477-020-0733-0).\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/GenomePubs.png\"\n",
    "    alt=\"Pangenome Publications\" />\n",
    "  <figcaption><a href=\"https://www.nature.com/articles/s41477-020-0733-0\">https://www.nature.com/articles/s41477-020-0733-0</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "### \"Pangenome\" Today\n",
    "\n",
    "Today, the definition of a pangenome has evolved:\n",
    "\n",
    "“Any collection of genomic sequences to be analyzed jointly or to be used as a reference. These sequences can be linked in a graph-like structure, or simply constitute sets of (aligned or unaligned) sequences.” – [Computational Pangenomics Consortium](https://academic.oup.com/bib/article/19/1/118/2566735)\n",
    "\n",
    "### The Benefit of Pangenomes\n",
    "\n",
    "+ Removes reference bias that:\n",
    "  + May only represent one organism\n",
    "  + Could be a “mosaic”of individuals, i.e. doesn’t represent a coherent haplotype\n",
    "  + May contain allele bias\n",
    "  + Doesn’t include common variation\n",
    "+ Allows multiple assemblies to be analyzed simultaneously, i.e. efficiently\n",
    "\n",
    "###  What are pangenomes good for?\n",
    "\n",
    "+ Core vs dispensable genes:\n",
    "  + How big is the core?\n",
    "  + How big is the dispensable?\n",
    "  + How big is the pangenome?\n",
    "  + What traits are associated with the core/dispensable?\n",
    "+ Unbiased read mapping and variant calling\n",
    "+ More robust variation-trait association\n",
    "+ Visual exploration of genomic structure of population\n",
    "\n",
    "----------------\n",
    "\n",
    "##  Computational Pangenomics\n",
    "\n",
    "“Questions about efficient data structures, algorithms and statistical methods to perform bioinformatic analyses of pan-genomes give rise to the discipline of ‘computational pan-genomics’.” – [Computational Pangenomics Consortium](https://academic.oup.com/bib/article/19/1/118/2566735)\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/Computational.png\"\n",
    "    alt=\"Computational Pangenomics\" />\n",
    "  <figcaption><a href=\"https://academic.oup.com/bib/article/19/1/118/2566735\">https://academic.oup.com/bib/article/19/1/118/2566735</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "### Pangenome Representations\n",
    "\n",
    "Pangenomes can be represented in many different formats:\n",
    "\n",
    "+ Gene sets\n",
    "+ Multiple sequence alignments\n",
    "+ K-mer sets\n",
    "+ Graphs\n",
    "  + De Bruijn graphs\n",
    "  + Haptotype graphs\n",
    "  + Variation graphs\n",
    "\n",
    "\n",
    "### Variation Graphs\n",
    "\n",
    "We will focus on variation graphs, which most pangenomic tools use. These graphs consist of nodes that contain sequence data, edges that connect the nodes, and paths that thread genomes, chromosomes, haplotypes, genes, or other sequence information through the graph, thereby connecting original sequence inputs directly with walks through the graph. Variation graphs retain all variation present in the original input sequences and balance construction efficiency with ease of visualization and interpretation.\n",
    "\n",
    "+ Variation forms bubbles\n",
    "+ Nodes represent sequences\n",
    "+ Chains of nodes represent contiguous sequence in one or more assemblies\n",
    "+ The sequences of nodes connected by an edge may overlap\n",
    "+ Graphs can be acyclic or cyclic\n",
    "+ Haplotypes are “threaded” through graph as paths\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/VariationGraph.jpeg\"\n",
    "    alt=\"Pangenome Representations\" />\n",
    "  <figcaption><a href=\"https://academic.oup.com/bib/article/19/1/118/2566735\">https://academic.oup.com/bib/article/19/1/118/2566735</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "### Types of Variation Graphs\n",
    "\n",
    "1. Reference Graph (vg)\n",
    "      + A reference with variants\n",
    "      + e.g., [Human reference now includes VCF with common variation](https://www.ncbi.nlm.nih.gov/genome/guide/human/)\n",
    "2. Reference Backbone; “iterative” (minigraph)\n",
    "      + Graph starts as reference and other sequences are layered on, i.e. variants can be relative to sequences other than the reference\n",
    "3. Reference-Free (Cactus and pggb)\n",
    "      + Graph is built using non-reference techniques, such as multiple sequence alignment\n",
    "\n",
    "These are all methods used by the [Human Pangenome Reference Consortium](https://humanpangenome.org). In this module, you will learn how to use pggb, which aims to capture all the variation present in the input sequences and does not designate a genome as the reference, thereby avoiding reference bias.\n",
    "\n",
    "### Mapping Reads to Variation Graphs\n",
    "\n",
    "You can map sequencing reads from an individual to variation graphs, splitting the reads across matchin nodes as shown below [(Hickey et al, 2020)](https://link.springer.com/article/10.1186/s13059-020-1941-7). This allows you to identify which of the nodes (and, therefore, which variants) are supported by the reads. This read support can then be converted into genotypes for individual you sequenced and aligned to the variation graph.\n",
    "\n",
    "<figure>\n",
    "  <img\n",
    "    src=\"./Figures/ReadsToVariation.png\"\n",
    "    alt=\"Genotyping Variation\" />\n",
    "  <figcaption><a href=\"https://link.springer.com/article/10.1186/s13059-020-1941-7\">https://link.springer.com/article/10.1186/s13059-020-1941-7</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "----------------\n",
    "\n",
    "## Pangenome Data Sets\n",
    "\n",
    "Here are some links to some pangenome data sets you can explore.\n",
    "\n",
    "+ [Human Reference + Variation VCF](https://www.ncbi.nlm.nih.gov/genome/guide/human/)\n",
    "+ [Human Pangenome Reference Consortium](https://humanpangenome.org)\n",
    "+ [Zoonomia (200 mammals) Project](https://zoonomiaproject.org/the-data/)\n",
    "+ [Maize NAM founder genomes](https://www.science.org/doi/10.1126/science.abg5289)\n",
    "+ [Yeast Population Reference Panel (YPRP)](https://yjx1217.github.io/Yeast_PacBio_2016/welcome/)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------------------\n",
    "\n",
    "### Quiz\n",
    "\n",
    "Run the code below to take the quiz."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame('../html/quiz_pangenomes.html', width=800, height=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------------------\n",
    "\n",
    "## Conclusion\n",
    "This module gave an overview of pangenomes and the different types of pangenome representations, focusing on variation graphs. In the next submodule, you will build some pangenome graphs.\n",
    "\n",
    "----------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean up\n",
    "\n",
    "<div class=\"alert alert-warning\">No cleanup is necessary for this submodule. Don't forget to shutdown your Workbench when you are done working through this module!.</div>"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-env-nigms-pangenomics-nigms-pangenomics",
   "name": "workbench-notebooks.m127",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m127"
  },
  "kernelspec": {
   "display_name": "nigms-pangenomics",
   "language": "python",
   "name": "conda-env-nigms-pangenomics-nigms-pangenomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
